base_layer/wallet/src/transaction_service/service.rs

```diff
@@ -631,6 +651,101 @@ where
         Ok(())
     }
 
+    /// Request a tx_id and spending_key for a coinbase output to be mined
+    pub async fn request_coinbase_key(
+        &mut self,
+        amount: MicroTari,
+        maturity_height: u64,
+    ) -> Result<PendingCoinbaseSpendingKey, TransactionServiceError>
+    {
+        let mut rng = rand::OsRng::new().unwrap();
+        let tx_id: TxId = rng.next_u64();
+
+        let spending_key = self
+            .output_manager_service
+            .get_coinbase_spending_key(tx_id.clone(), amount.clone(), maturity_height)
+            .await?;
+
+        self.db
+            .add_pending_coinbase_transaction(tx_id.clone(), PendingCoinbaseTransaction {
+                tx_id,
+                amount,
+                commitment: self
+                    .factories
+                    .commitment
+                    .commit_value(&spending_key, u64::from(amount.clone())),
+                timestamp: Utc::now().naive_utc(),
+            })?;
+
+        Ok(PendingCoinbaseSpendingKey { tx_id, spending_key })
+    }
+
+    /// Once the miner has constructed the completed Coinbase transaction they will submit it to the Transaction Service
+    /// which will monitor the chain to see when it has been mined.
+    pub fn submit_completed_coinbase_transaction(
+        &mut self,
+        tx_id: TxId,
+        completed_transaction: Transaction,
+    ) -> Result<(), TransactionServiceError>
+    {
+        let coinbase_tx = self.db.get_pending_coinbase_transaction(tx_id.clone()).map_err(|e| {
+            error!(
+                target: LOG_TARGET,
+                "Finalized coinbase transaction TxId does not exist in Pending Inbound Transactions"
+            );
+            e
+        })?;
+
+        if completed_transaction.body.inputs().len() != 0 || completed_transaction.body.outputs().len() != 1 {
+            error!(
+                target: LOG_TARGET,
+                "Provided Completed Transaction for Coinbase Transaction does not contain just a single output"
+            );
+            return Err(TransactionServiceError::InvalidCompletedTransaction);
+        }
+
+        if coinbase_tx.commitment != completed_transaction.body.outputs()[0].commitment ||
+            completed_transaction.body.outputs()[0].features.flags != OutputFlags::COINBASE_OUTPUT
+        {
+            error!(
+                target: LOG_TARGET,
+                "Provided Completed Transaction commitment for Coinbase Transaction does not match the stored \
+                 commitment for this TxId"
+            );
+            return Err(TransactionServiceError::InvalidCompletedTransaction);
+        }
+
```

Comment:
I think it will be good to check the flags as well. The output and kernel should have a coinbase flag set.

Reply:
I do check the flag in the If statement above?

Reply:
oops, missed that one. 
But the kernel should be marked as well. Not just the output. It also has a flag for coinbase which is set.

Reply:
done